holosqape/socketinterface.cpp

```diff
@@ -0,0 +1,121 @@
+    //QObject::connect(&m_webSocket, QOverload<const QList<QSslError>&>::of(&QWebSocket::sslErrors),
+              //this, &SslEchoClient::onSslErrors);
+}
+
+void SocketInterface::setContainer(Container* container) {
+    m_container = container;
+}
+
+
+void SocketInterface::incomingConnection() {
+    QWebSocket* socket = m_server.nextPendingConnection();
+    if(!socket) return;
+
+    qDebug() << "New websocket connection";
+
+    m_web_sockets << socket;
+    QObject::connect(socket, &QWebSocket::textMessageReceived,
+                     this, &SocketInterface::message_received
+                     );
+}
+
+void SocketInterface::message_received(const QString &message) {
+    QWebSocket* socket = qobject_cast<QWebSocket*>(sender());
+
+
+    QJsonDocument doc = QJsonDocument::fromJson(message.toUtf8());
+
+    if(!doc.isObject()) {
+        qDebug() << "Received RPC that is not an object:";
+        qDebug() << message;
+        socket->sendTextMessage(QString('{"error": "receveid non-JSON-object"}'));
+        return;
+    }
+
+    QJsonObject rpc = doc.object();
+
+    if(!rpc.contains("id")) {
+        qDebug() << "Received RPC without id:";
+        qDebug() << message;
+        socket->sendTextMessage(QString("{\"error\": \"missing id\"}"));
+        return;
+    }
+
+    if(!rpc.contains("method")) {
+        qDebug() << "Received RPC without method:";
+        qDebug() << message;
+        socket->sendTextMessage(QString("{\"error\": \"missing method\", \"id\": %1}").arg(rpc.value("id").toInt()));
+        return;
+    }
+
+    if(!rpc.contains("params")) {
+        qDebug() << "Received RPC without params:";
+        qDebug() << message;
+        socket->sendTextMessage(QString("{\"error\": \"missing params\", \"id\": %1}").arg(rpc.value("id").toInt()));
+        return;
+    }
+
+    qDebug() << "RPC Request: " << message;
+
+    QJsonObject result = executeRPC(rpc);
+    QString string_result = QJsonDocument(result).toJson();
+
+    socket->sendTextMessage(string_result);
+
+    qDebug() << "RPC Result: " << string_result;
+}
+
+QJsonObject SocketInterface::executeRPC(QJsonObject rpc) {
+    QJsonObject result;
+    result["id"] = rpc["id"].toInt();
+
+    QStringList method = rpc.value("method").toString().split("/");
+    if(method.length() == 4) {
+        QString dna = method[0];
+        QString zome = method[1];
+        QString cap = method[2];
+        QString function = method[3];
+        QString params = QJsonDocument(rpc.value("params").toObject()).toJson();
+
+        App* app = 0;
+        foreach(App* a, m_container->instances()){
+            if(a->hash() == dna)
+                app = a;
+        }
+
+        if(!app) {
+            qDebug() << "app == 0";
+            result["error"] = QString("could not find running instance with hash %1").arg(dna);
+            return result;
+        }
+
+        QString result_string = app->call(zome, cap, function, params);
```

Comment:
I don't know how error handling works in this code, are any exceptions possible?

Reply:
Only unexpected panics..